### PR TITLE
Add basic support for nonzero motion vectors

### PR DESCRIPTION
**Pull request by fbossen**
Add basic support for prediction and coding of nonzero vectors without doing actual motion estimation

### CODE REVIEW

**Comment by ycho** [comment]
FYI, av1_encode_mv() is available at: https://github.com/xiph/rav1e/pull/446
(Sorry- not tested)

**Comment by fbossen** [comment]
@ycho Thanks for reminding me. Not quite there yet...

File: src/partition.rs [hunk 1]
```diff
@@ -511,22 +518,33 @@ impl PredictionMode {
   }
 
   pub fn predict_inter<'a>(self, fi: &FrameInvariants, p: usize, po: &PlaneOffset,
-                           dst: &'a mut PlaneMutSlice<'a>, plane_size: BlockSize) {
+                           dst: &'a mut PlaneMutSlice<'a>, plane_size: BlockSize,
```
Comment (ycho) [inline src/partition.rs]: How about saying plane_**b**size for plane_size?

**Comment by ycho** [comment]
master-48326cdc05f9f042ae05e6d7066f1b89152daaba -> fbossen_mc4_rebased

   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000
-2.4781 | -2.9127 | -3.3728 |  -2.7879 | -3.1224 | -3.9666 |    -3.1537

**Comment by fbossen** [comment]
Maybe should do something better than random MVs

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.2129 | -1.7391 | -1.8642 |  -1.2820 | -1.4119 | -1.4238 |    -1.5448 |

**Comment by fbossen** [comment]
It's better with some really basic ME

|     PSNR |  PSNR Cb |  PSNR Cr | PSNR HVS |     SSIM |  MS SSIM | CIEDE 2000 |
|     ---: |     ---: |     ---: |     ---: |     ---: |     ---: |       ---: |
| -21.6789 | -24.1395 | -24.8527 | -22.3152 | -23.4495 | -23.6358 |   -23.5536 |

**Comment by ycho** [comment]
Wow!

**Comment by tmatth** [comment]
Holy cow

File: src/me.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+use plane::PlaneOffset;
+use context::BLOCK_TO_PLANE_SHIFT;
+
+pub fn motion_estimation(fi: &FrameInvariants, fs: &mut FrameState, bsize: BlockSize,
+                         bo: &BlockOffset, ref_frame: usize) -> MotionVector {
+
+  match fi.rec_buffer.frames[fi.ref_frames[ref_frame - LAST_FRAME]] {
+    Some(ref rec) => {
+      let po = PlaneOffset { x: bo.x << BLOCK_TO_PLANE_SHIFT, y: bo.y << BLOCK_TO_PLANE_SHIFT };
+      let range = 16 as usize;
```
[the remark this answers is not in the document]
Comment (fbossen) [reply src/me.rs]: Yes. Note that only vectors that are multiples of 2.0 are searched to avoid having to do any filtering (including for chroma)

**Comment by ycho** [comment]
Oh, step by 2, thats why it is so fast... :)
Btw, locally tested for -s 0 and 1 on ubuntu and caused NO issues.
Reading the code now....

File: src/encoder.rs [hunk 3]
```diff
@@ -1146,7 +1146,19 @@ pub fn encode_block_b(fi: &FrameInvariants, fs: &mut FrameState,
             //let mode_context = if bo.x == 0 && bo.y == 0 { 0 } else if bo.x ==0 || bo.y == 0 { 51 } else { 85 };
             // NOTE: Until rav1e supports other inter modes than GLOBALMV
```
Comment (ycho) [inline src/encoder.rs]: Later, we remove this obsolete comment line.

Comment (fbossen) [reply src/encoder.rs]: yes indeed

**Comment by ycho** [comment]
LGTM. Are you going to squash into ONE commit?

**Comment by fbossen** [comment]
@ycho Yes, the intent is to squash into one commit

**Pull request merged**